File: HadoopDefinitiveGuide/src/main/scala/ch19/MaxTemperature.scala
package ch19

import org.apache.spark.{SparkConf, SparkContext}

object MaxTemperature {
	def main(args: Array[String]): Unit = {
		val conf = new SparkConf().setAppName("MaxTemperature").setMaster("local")
		val sc = new SparkContext(conf)

		sc.textFile("e:/temp/sample.txt")
			.map(_.split("\t"))
			.filter(rec => (rec(1) != "9999" && rec(2).matches("[01459]")))
			.map(rec => (rec(0).toInt, rec(1).toInt))
			.reduceByKey((a, b) => Math.max(a, b))
			.foreach(println(_))


	}
}

File: HadoopDefinitiveGuide/src/main/java/ch19/MaxTemperature.py
# -*- coding:utf-8 -*-

from pyspark import SparkContext
import re, sys

sc = SparkContext("local", "Max Temperature")
sc.textFile("e:/temp/sample.txt") \
    .map(lambda s: s.split("\t") \
    .filter(lambda rec: (rec[1] != "9999" and re.match("[01459]", rec[2]))) \
    .map(lambda rec: (int(rec[0]), int(rec[1]))) \
    .reduceByKey(max) \
    .saveAsTextFile("e/temp/result.txt")